common/persistence/cassandra/cassandraPersistence.go

```diff
@@ -1078,11 +1089,77 @@ func (d *cassandraPersistence) GetShard(
 		}
 	}
 
-	info := createShardInfo(d.currentClusterName, result["shard"].(map[string]interface{}))
+	// check if rangeID column and rangeID field in shard column matches, if not we need to pick the larger
+	// rangeID.
+	//
+	// If shardInfoRangeID < rangeID, we don't need to do anything here as createShardInfo will ignore
+	// shardInfoRangeID and return rangeID instead. Later when updating the shard, CAS can still succeed
+	// as the value from rangeID columns is returned, shardInfoRangeID will also be updated to the correct value.
+	rangeID := result["range_id"].(int64)
+	shard := result["shard"].(map[string]interface{})
+	shardInfoRangeID := shard["range_id"].(int64)
+	if shardInfoRangeID > rangeID {
+		// In this case we need to fix the rangeID column before returning the result as:
+		// 1. if we return shardInfoRangeID, then later shard CAS operation will fail
+		// 2. if we still return rangeID, CAS will work but rangeID will move backward which
+		// result in lost tasks, corrupted workflow history, etc.
```

Comment:
Maybe add a fatal logging there

Reply:
We don't need to panic here, the logic below can fix the corrupted rangeID. I can add a Warning log though.